cladecombiner/aggregator.py

```diff
@@ -0,0 +1,228 @@
+from warnings import warn
+
+from .taxon import Taxon
+from .taxon_utils import sort_taxa
+from .taxonomy_scheme import PhylogeneticTaxonomyScheme
+from .utils import table
+
+
+class Aggregation(dict[Taxon, Taxon]):
+    """
+    An object for aggregations, basically just a dictionary.
+    """
+
+    def _validate(
+        self, input_taxa: Iterable[Taxon], taxon_map: dict[Taxon, Taxon]
+    ):
+        """
+        Checks that all input taxa have been mapped exactly once.
+        """
+        if set(taxon_map.keys()) != set(input_taxa):
+            raise RuntimeError(
+                "Mismatch between aggregated taxa and input taxa. Input taxa are: "
+                + str(input_taxa)
+                + " but aggregated taxa are "
+                + str(taxon_map.keys())
+            )
+        tab = table(taxon_map)
+        if not all(v == 1 for v in tab.values()):
+            raise RuntimeError(
+                "Found following taxa mapped more than once: "
+                + str([k for k, v in tab.items() if v > 1])
+            )
+
+    def __init__(
+        self, input_taxa: Iterable[Taxon], taxon_map: dict[Taxon, Taxon]
+    ):
+        self._validate(input_taxa, taxon_map)
+        super().__init__(taxon_map)
+
+    def to_str(self):
+        """
+        Get str : str map of taxa names
+        """
+        return {k.name: v.name for k, v in self.items()}
+
+
+class Aggregator(ABC):
+    """
+    Aggregators return Aggregations, maps of input_taxon : aggregated_taxon
+    """
+
+    @abstractmethod
+    def aggregate(self, input_taxa: Iterable[Taxon]) -> Aggregation:
+        pass
+
+
+class ArbitraryAggregator(Aggregator):
+    """
+    Aggregation via a user-provided dictionary.
+    """
+
+    def __init__(
+        self,
+        map: dict[Taxon, Taxon],
+    ):
+        """
+        FixedAggregator constructor.
+
+        Parameters
+        ----------
+        map : dict[Taxon, Taxon]
+            Dictionary mapping the input taxa to their aggregated taxa.
+        """
+        self.map = map
+
+    def aggregate(self, input_taxa: Iterable[Taxon]) -> Aggregation:
+        return Aggregation(
+            input_taxa, {taxon: self.map[taxon] for taxon in input_taxa}
+        )
+
+
+class BasicPhylogeneticAggregator(Aggregator):
+    """
+    An aggregator which maps a set of input taxa to a fixed set of aggregation targets using a tree.
+    """
+
+    def __init__(
+        self,
+        targets: Iterable[Taxon],
+        taxonomy_scheme: PhylogeneticTaxonomyScheme,
+        sort_clades: bool = True,
+        unmapped_are_other: bool = True,
+        warn: bool = True,
+    ):
+        """
+        BasicPhylogeneticAggregator constructor.
+
+        Parameters
+        ----------
+        targets : Iterable[Taxon]
+            The taxa into which we wish to aggregate the input taxa.
+
+        taxonomy_scheme : PhylogeneticTaxonomyScheme
+            The tree which we use to do the mapping.
+
+        sort_clades : bool
```

Comment:
Should this be left to the user to do via serial aggregation?

Eg I think it's much clearer to implement "aggregate to BA.2 except BA.2.86" as "aggregate to BA.2.86, then aggregate to BA.2". That would also make it easier for the user to interrogate what's going on at each step.

Although I guess this would require a new argument, which is taxa that must map to themselves. So you'd need to do:
```
SerialAggregator([
  PhyloAggregator(targets=[BA_2_86]),
  PhyloAggregator(targets=[BA_2], do_not_further_remap=[BA_2_86]
])
```
which is kind of gross

Are there any cases where one might plausibly want to do some kind of "except" mapping that can't be handled by serially aggregating within this aggregator, rather than needing to use `SerialAggregator` and some kind of "do not remap" thing?

Reply:
I think there is an argument to be made that, as-is, this is a combination of an "atomic" Aggregator that could be used in `SerialAggregator`s and a higher-level pre-baked class implementing a pre-set `SerialAggregator`. The cleaning up bit of self-aggregation or putting them in "other" is the latter, the main loop in `.aggregate()` the former.

I'm inclined to believe, at the moment, that it's validly "atomic" of this to take in a list of taxa and handle ordering. However, we will see what happens when we start trying to mix and match this with the to-be-implemented automatic aggregators, where such "except" statements will also crop up ("aggregate automatically, except I want these taxa un-aggregated...").